### PR TITLE
Adding TrimExcess Implementation to Dictionary

### PR DESCRIPTION
**Pull request by maryamariyan**
Issue is: https://github.com/dotnet/corefx/issues/24445
Tests and the ref changes are made in a corefx PR separately here dotnet/corefx#26239.

cc: @danmosemsft @benaadams 

TODO Next in a separate PR:
- Will add EnsureCapacity to SortedSet and HashSet
- Will add TrimExcess to SortedSet

### CODE REVIEW

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 1]
```diff
@@ -777,6 +775,67 @@ public int EnsureCapacity(int capacity)
...
+                        ref Entry entry = ref _entries[i];
+                        int bucket = entry.hashCode % capacity;
+                        entry.next = buckets[bucket];
+                        buckets[bucket] = i;
+                        entries[k] = entry;
+                        k++;
+                    }
+                }
+
+                Array.Clear(_entries, 0, count);
```
Comment (benaadams) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Is needed? `_entries` is just being dropped and GC'd

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 2]
```diff
@@ -777,6 +775,66 @@ public int EnsureCapacity(int capacity)
             return newSize;
         }
 
+        /// <summary>
+        /// Sets the capacity of this dictionary to hold up at least all existing entries
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: This is not strictly true is it? It picks a larger prime.
How about "Sets the capacity of this dictionary to what it would be if it had been originally initialized with all its entries"

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 3]
```diff
@@ -777,6 +775,66 @@ public int EnsureCapacity(int capacity)
...
+        }
+
+        /// <summary>
+        /// Sets the capacity of this dictionary to hold up at least 'capacity' entries without any further expansion of its backing storage
+        /// </summary>
+        public void TrimExcess(int capacity)
+        {
+            if (capacity < Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: We should try to start including the invalid value in argument exceptions so that callstacks are more actionable. AOORE already has a constructor `public ArgumentOutOfRangeException(String paramName, Object actualValue, String message)` that we don't seem to use. If you want you could add an overload in ThrowHelper so that you could call this one so you can pass in the value of capacity.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 4]
```diff
@@ -777,6 +775,66 @@ public int EnsureCapacity(int capacity)
...
+            }
+
+            if (_entries != null && capacity < _entries.Length)
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: @jkotas from a CQ perspective, as a rule of thumb is early return better, worse, or not generally one way or another?

That aside, I generally prefer to early out when the rest of the method would be enclosed by the if, as it's easier to follow and indents less.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 5]
```diff
@@ -777,6 +775,66 @@ public int EnsureCapacity(int capacity)
...
+                        buckets[bucket] = i;
+                        entries[k] = entry;
+                        k++;
+                    }
+                }
+                
+                _freeList = -1;
+                _freeCount = 0;
+                _version++;
+                _count = k;
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: `_count` is already correct isn't it? no need to set it. you can Debug.Assert _count == k if you want.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 6]
```diff
@@ -777,6 +775,66 @@ public int EnsureCapacity(int capacity)
...
+        {
+            if (capacity < Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
+            }
+
+            if (_entries != null && capacity < _entries.Length)
+            {
+                int[] buckets = new int[capacity];
+                for (int i = 0; i < buckets.Length; i++)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Can't this just call `Resize`? I do not think we want to have another copy of the rehashing look here.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 7]
```diff
@@ -777,6 +775,66 @@ public int EnsureCapacity(int capacity)
...
+        /// 
+        /// To completely clear a dictionary and release all memory 
+        /// referenced by the dictionary, execute the following statements:
+        /// 
+        /// dictionary.Clear();
+        /// dictionary.TrimExcess();
+        /// </summary>
+        public void TrimExcess()
+        {
+            TrimExcess(HashHelpers.GetPrime(Count));
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Did we decide whether to GetPrime or use Count directly? I guess GetPrime will give better performance, so as you've written it, this overload doesn't lead to the "pit of failure"

cc @benaadams

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 6 above)
Comment (maryamariyan) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: It's a bit different than Resize. I can extract the similarities into a private method which both call though.

please notice that in Resize we are not clearing _freeCount and _freeList

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 5 above)
Comment (maryamariyan) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Since ``` Count = _count - _freeCount``` and in TrimExcess we are clearing _freeCount then _count does get changed

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 4 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: It should not matter in this case. I think that the JIT will generate identical code for it because of we have started folding the epilogs recently.

There is a much more fundamental code size issue in this delta in that it introduces another copy of the rehashing loop.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 8]
```diff
@@ -777,6 +775,66 @@ public int EnsureCapacity(int capacity)
...
+        public void TrimExcess(int capacity)
+        {
+            if (capacity < Count)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
+            }
+
+            if (_entries != null && capacity < _entries.Length)
+            {
+                int[] buckets = new int[capacity];
```
Comment (jkotas) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: The capacity needs to be prime. You need to round it up to the next prime.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 6 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: That's because of you have bugs in your implementation - you need to round up the capacity to the next prime, etc. Once you fix the bugs, they should be exactly same.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 5 above)
Comment (maryamariyan) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Please note, when we call Remove on Dictionary _count remains the same value but _freeCount gets incremented. When TrimExcess gets called and _freeCount would get cleared to zero.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 6 above)
Comment (benaadams) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: `TrimExcess` has to compress the `Dictionary` but `Resize` doesn't.
Calling `Resize` would; currently, also reallocate the arrays. Although when its rehashing its not actually adjusting the arrays so it doesn't need to reallocate - so that could be changed

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 7 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: It has to be GetPrime for correctness. If it is not a prime, you can end up with infinite loops.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 6 above)
Comment (maryamariyan) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: @jkotas I'm addressing getting next prime number as capacity in my next PR.
But the other difference between Resize and TrimExcess I am referring to, is the fact that when we resize, we don't necessarily clear _freeCount, but in TrimExcess we do.

Comment (jkotas) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: > I'm addressing getting next prime number as capacity in my next PR.

I assume that you meant next commit. It does not make sense to delay it to next PR.

Once you start rounding up the capacity to the next prime, you want to be setting `_freeCount` same way as `Resize` because you will have free entries at the end. You want those free entries to be available for future inserts.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 7 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Ah yes it's just missing below.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 5 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Ah yes.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 6 above)
Comment (benaadams) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Alternative would be to go down the freelist and move items from the end into the free spots; then call resize

Comment (danmosemsft) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: @jkotas re reusing Resize(), it copies all the entry array that has entries (whether free or not) using `            Array.Copy(_entries, 0, entries, 0, count);`. That may be the entire existing Entry array. That would not fit in the smaller array that TrimExcess wants. 

I think it is necessary to do what @benaadams proposes, to defrag all the free entries to the end.

Could you clarify, if I misunderstand?

Comment (jkotas) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Hmmm.

What would be the impact of replacing the `Array.Copy` with compacting in all cases? It may actually be faster for large dictionaries in some cases because of we would be faulting the array into the cache just once, not twice like with the current code.

The Dictionary code gets stamped many times, especially with AOT. When we are trying to improve Dictionary, we need to spend the extra time to minimize this impact.

Sooner or later, we will need to start the "Power Collections" library where we can go wild with different optimizations, implement collections tailored for different use cases, without worrying about the Dictionary constrains...

Comment (danmosemsft) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Compacting would also end up with an Entries array that has potentially better locality. Do you want @maryamariyan to pause this change, and prototype/measure such a change to Resize() ?

Comment (jkotas) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Yes, I think we should evaluate different options.

Comment (maryamariyan) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Thanks. I'll address the other comments in the meantime.

**Comment by maryamariyan** [comment]
This report is checking the time it takes to Resize dictionary to a smaller size.
**Pre** is the existing Resize Implementation
**Post** is the updated Resize Implementation from this PR.

| Type | Use Case | millisecs | start capacity | initial size | updated capacity | new size | resize to |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Post | A | 70.44ms | 42046 | 10000 | 43627 | 10000 | 21023 |
| Pre | A | 70.82ms | 42046 | 10000 | 43627 | 10000 | 21023 |
| Post | A | 830.43ms | 450614 | 100000 | 467237 | 100000 | 225307 |
| Pre | A | 854.69ms | 450614 | 100000 | 467237 | 100000 | 225307 |
| Post | B | 33.11ms | 42046 | 10000 | 43627 | 5000 | 21023 |
| Pre | B | 33.20ms | 42046 | 10000 | 43627 | 5000 | 21023 |
| Post | B | 392.85ms | 450614 | 100000 | 467237 | 50000 | 225307 |
| Pre | B | 417.53ms | 450614 | 100000 | 467237 | 50000 | 225307 |
| Post | C | 78.42ms | 10000 | 21023 | 75431 | 11023 | 42046 |
| Pre | C | 78.92ms | 10000 | 21023 | 75431 | 11023 | 42046 |
| Post | C | 1059.76ms | 100000 | 225307 | 807403 | 125307 | 450614 |
| Pre | C | 1086.05ms | 100000 | 225307 | 807403 | 125307 | 450614 |


Use Case A:
- Dictionary is full

Use Case B:
- Setup stage: Initial capacity is high, we do some add and then remove to dictionary

Use Case C:
- Setup stage: Initial capacity is low, after add and remove the capacity has increased.

**Comment by jkotas** [comment]
> This report shows the time it takes to Resize dictionary to a smaller size.

What does this change do to Resizing dictionary to larger size? 

Resizing dictionary to smaller size is rare operation. It is no big deal if it is a bit slower. Resizing dictionary to larger is much more common operation.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 9]
```diff
@@ -535,16 +532,25 @@ private void Resize(int newSize, bool forceNewHashCodes)
...
 
+            int k = 0;
             for (int i = 0; i < count; i++)
             {
-                if (entries[i].hashCode >= 0)
+                if (_entries[i].hashCode >= 0)
                 {
-                    int bucket = entries[i].hashCode % newSize;
-                    entries[i].next = buckets[bucket];
+                    ref Entry entry = ref _entries[i];
```
Comment (jkotas) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Caching ` ref Entry entry = ref _entries[i];` in a local before the if-check and using it in the if-check would be better, I think.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 10]
```diff
@@ -535,16 +532,25 @@ private void Resize(int newSize, bool forceNewHashCodes)
                 }
             }
 
+            int k = 0;
             for (int i = 0; i < count; i++)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Does the loop need to be to `_entries.Count` now?

Comment (jkotas) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: It may be profitable to cache `_entries` in a local. The JIT should be able to eliminate the bounds check that way:

```
Entry[] oldEntries = _entries;
for (int i = 0; i < oldEntries.Count; i++)
```

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 9 above)
Comment (mikedn) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: The JIT tends to handle well repeated accesses to the same array element so caching element refs may not do much. What might help in this this kind of code is caching the array itself (`_entries`) in a local.

**Comment by maryamariyan** [comment]
| Type | Use Case | millisecs | start capacity | initial size | updated capacity | new size | resize to |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Post | A | 98.07ms | 42046 | 10000 | 43627 | 10000 | 63069 |
| Pre | A | 97.04ms | 42046 | 10000 | 43627 | 10000 | 21023 |
| Post | A | 1006.10ms | 450614 | 100000 | 467237 | 100000 | 675921 |
| Pre | A | 1012.50ms | 450614 | 100000 | 467237 | 100000 | 675921 |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Post | B | 49.43ms | 42046 | 10000 | 43627 | 5000 | 63069 |
| Pre | B | 47.86ms | 42046 | 10000 | 43627 | 5000 | 63069 |
| Post | B | 506.94ms | 450614 | 100000 | 467237 | 50000 | 675921 |
| Pre | B | 508.18ms | 450614 | 100000 | 467237 | 50000 | 675921|
| --- | --- | --- | --- | --- | --- | --- | --- |
| Post | C | 106.83ms | 10000 | 21023 | 75431 | 11023 | 84092 |
| Pre | C | 106.85ms | 10000 | 21023 | 75431 | 11023 | 84092 |
| Post | C | 1293.48ms | 100000 | 225307 | 807403 | 125307 | 901228 |
| Pre | C | 1269.52ms | 100000 | 225307 | 807403 | 125307 | 901228|


Use Case A:
- Dictionary is full

Use Case B:
- Setup stage: Initial capacity is high, we do some add and then remove to dictionary

Use Case C:
- Setup stage: Initial capacity is low, after add and remove the capacity has increased.

This report shows the time it takes to Resize dictionary to a **larger size*** this time.
**Pre** is the existing Resize Implementation
**Post** is the updated Resize Implementation from this PR.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs (hunk 9 above)
Comment (maryamariyan) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: I will push up another commit shortly to address this comment

**Comment by maryamariyan** [comment]
| Type | Use Case | millisecs | start capacity | initial size | updated capacity | new size | resize to |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Post | B | 10.006ms | 42046 | 10000 | 43627 | 1000 | 21023 |
| Pre | B | 9.970ms | 42046 | 10000 | 43627 | 1000 | 21023 |
| Post | B | 100.653ms | 450614 | 100000 | 467237 | 10000 | 225307 |
| Pre | B | 101.261ms | 450614 | 100000 | 467237 | 10000 | 225307 |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Post | B | 47.936ms | 42046 | 10000 | 43627 | 5000 | 21023 |
| Pre | B | 48.075ms | 42046 | 10000 | 43627 | 5000 | 21023 |
| Post | B | 509.139ms | 450614 | 100000 | 467237 | 50000 | 225307 |
| Pre | B | 512.562ms | 450614 | 100000 | 467237 | 50000 | 225307 |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Post | B | 89.538ms | 42046 | 10000 | 43627 | 9000 | 21023 |
| Pre | B | 89.765ms | 42046 | 10000 | 43627 | 9000 | 21023 |
| Post | B | 923.998ms | 450614 | 100000 | 467237 | 90000 | 225307 |
| Pre | B | 925.274ms | 450614 | 100000 | 467237 | 90000 | 225307 |

Use Case B:
- Setup stage: Initial capacity is high, we do some add and then remove to dictionary

@danmosemsft please note, 
- This chart shows the time it takes for Use Case B to Resize dictionary to a **smaller size**.
- Three different variation of zombies are used: (1) with 10% zombie entries, (2) with 50% zombies entries, (3) with 90% zombie entries

**Pre** is the existing Resize Implementation
**Post** is the updated Resize Implementation from this PR.

**Comment by maryamariyan** [comment]
Actually, please hold off on looking too much into the charts.
I am currently in the process of updating the charts to make them clearer, and explaining further what the use cases do, and adding conclusion.

I am also going to add the error and standard deviation next to the timings.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 11]
```diff
@@ -777,6 +783,42 @@ public int EnsureCapacity(int capacity)
             return newSize;
         }
 
+        /// <summary>
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: you'll want to make the comments on the two overloads consistent with each other

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 12]
```diff
@@ -535,16 +532,25 @@ private void Resize(int newSize, bool forceNewHashCodes)
...
-                if (entries[i].hashCode >= 0)
+                if (_entries[i].hashCode >= 0)
                 {
-                    int bucket = entries[i].hashCode % newSize;
-                    entries[i].next = buckets[bucket];
+                    ref Entry entry = ref _entries[i];
+                    int bucket = entry.hashCode % newSize;
+                    entry.next = buckets[bucket];
                     buckets[bucket] = i;
+                    entries[k] = entry;
```
Comment (danmosemsft) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: should this be 
```c#
buckets[bucket] = k;
entries[k] = entry;
```
`i` is not relevant to the new arrays.

Comment (danmosemsft) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: That would break lookup so you might want to run the Dictionary tests against this change.

**Comment by maryamariyan** [comment]
Use Case E:

- Initialize dictionary with small capacity.
- Add as much elements to fill up the capacity
- Resize to HashHelpers.ExpandPrime(capacity). (Measurements above are done for this step only)

Please note I'm checking this use case, because I believe this can closely simulate the resize operation inside the Add operation of Dictionary.

| Type | Use Case | Mean | Error | StdDev | initial capacity | initial count | new capacity | new count | resize to (with measurement) |
| After | C | 2.717ms | 0.0600ms | 0.1479ms | 107 | 239 | 239 | 521 |
| Before | C | 2.736ms | 0.0546ms | 0.1026ms | 107 | 239 | 239 | 521 |
| After | C | 225.3ms | 4.932ms | 5.482ms | 10103 | 21023 | 21023 | 21023 | 43627  |
| Before | C | 192.1ms | 3.877ms | 5.683ms | 10103 | 21023 | 21023 | 21023 | 43627 |
| After | C | 773.3ms | 17.62ms | 22.29ms | 50000 | 108631 | 108631 | 108631 | 225307 |
| Before | C | 771.1ms | 12.44ms | 11.03ms | 50000 | 108631 | 108631 | 108631 | 225307 |

Conclusion: TBA

**Comment by maryamariyan** [comment]
I changed the approach for adding TrimExcess API. In this new approach I wouldn't change existing Resize API.

Note: ```_count``` is the sum of the number of used and unused entry elements in entries array. In other words it is the sum of unused entries (_freeCount) and dictionary.Count.

In this new approach, TrimExcess:
1 - modifies _freeList by removing unused entries in index range ```[newSize, _count)```
2 - moved the used entries in index range ```[newSize, _count)``` up into available unused entry location
3 - called resize to create new bucket and entries array of size newSize.

**Comment by jkotas** [comment]
Why is this better - is native code for this smaller?

**Comment by maryamariyan** [comment]
My older approach would have needed to update the Resize.
This way the performance of Resize on TryInsert (usecase D above) wouldn't be affected and the Resize logic wouldnt need to change.
Also I'm reusing some logic from Resize inside TryInsert.

**Comment by jkotas** [comment]
My primary concern with this change is how much (native) code this adds to each Dictionary instantiation.

**Comment by maryamariyan** [comment]
I removed the older benchmark values, and instead added a more recent ones below.

Use Case D:

- Initialize dictionary with small capacity.
- Add at least twice as many entries such that the dictionary becomes full, (is at capacity),
- Resize to HashHelpers.ExpandPrime(capacity). (Measurements above are done for this step only)

Please note I'm checking this use case, because I believe this can closely simulate the resize operation inside the Add operation of Dictionary.

| Use Case | Type | Mean | Error | initial capacity | initial count | new capacity | resize to (with measurement) |
| --- | --- | --- | --- | --- | --- | --- | --- |
| D | Before | 249.6ms | 3.218ms | 10103 | 21023 | 21023 | 43627 |
| D | Approach1 | 341.3ms | 3.685ms | 10103 | 21023 | 21023 | 43627  |
| D | Approach2 | 249.8ms | 3.063ms | 10103 | 21023 | 21023 | 43627  |

**Before** is the existing Resize Implementation, where we also copy over zombie elements in the dictionary
**Approach1** is the updated Resize Implementation shown in commit 08a5bf9 where we only copy over non-zombie elements into newly resized entry array.
**Approach2** is the updated Resize implementation shown in the latest commit (visible by looking at the entire PR diff). In this approach Resize stays the same as in master but we add TrimExcess

Conclusion:
Approach1 adds less to the native code (12Kb), compared to Approach2 (16Kb).
Approach1 performs slower for Use Case D, which happens on TryInsert anytime a resize is needed. Whereas Approach2 uses the same Resize method as in master branch. This explains why the above benchmark shows not much difference between "Approach2" and "Before" on Use Case D.

This comment explains my progress so far. I'll spend more time tomorrow figuring out if there is a better approach I can use.

**Comment by jkotas** [comment]
Thanks for collecting the size numbers. Approach1 vs. Approach2 - Approach2 looks better trade-off to me out of these two.

It may be also useful to get the number for your original impl.

**Comment by maryamariyan** [comment]
@jkotas the original approach adds 19Kb to native code. the implementation can be seen in commit 00c6b91

**Comment by maryamariyan** [comment]
@dotnet-bot test innerloop CentOS7.1 Release

**Comment by maryamariyan** [comment]
@dotnet-bot test help

**Comment by maryamariyan** [comment]
@dotnet-bot help

**Comment by jkotas** [comment]
Approach2  sounds the best so far.

**Comment by maryamariyan** [comment]
@dotnet/dnceng can you please take a look at the centos infrastructure issue?

**Comment by MattGal** [comment]
@maryamariyan looking now, thanks for bringing it to our attention.

**Comment by MattGal** [comment]
@maryamariyan it looks like infra is fine and there's an actual bug, and @jashook is already on top of it: https://github.com/dotnet/coreclr/pull/16035/files .  Definitely looks unrelated to your changes.  Once that merges, rebasing on it should allow a clean run, or you can make the judgment call of whether that  might actually have been broken by your changes.

**Comment by maryamariyan** [comment]
@jkotas I made the following correction to my comments above: the size increases I mentioned above were IL size increase amounts. (by mistake I said that's the amount that gets added to native code).

**Comment by maryamariyan** [comment]
@jkotas do you have further feedback for this PR? Do you approve Approach2 as is or should we investigate further for another option?

**Comment by jkotas** [comment]
> IL size increase amounts. (by mistake I said that's the amount that gets added to native code).

Are you sure? If you have measured System.Private.CoreLib.dll size in CoreCLR build, you have been measuring IL+native code size (the native code size would dominate the delta).

**Comment by maryamariyan** [comment]
What I did to get the numbers were to measure System.Private.CoreLib.dll size in CoreCLR.

**Comment by danmosemsft** [comment]
@jkotas then I misinformed @maryamariyan . This assembly includes crossgen native code?

**Comment by jkotas** [comment]
> This assembly includes crossgen native code?

Yes. CoreCLR build does crossgen by default. It is different from CoreFX build.

**Comment by danmosemsft** [comment]
Presumably however it would not give an indication of the number of instantiations - it would be necessary to attach a debugger to a "typical" app and use SOS.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 13]
```diff
@@ -777,6 +776,76 @@ public int EnsureCapacity(int capacity)
...
+        /// 
+        /// To completely clear a dictionary and release all memory 
+        /// referenced by the dictionary, execute the following statements:
+        /// 
+        /// dictionary.Clear();
+        /// dictionary.TrimExcess();
+        /// </summary>
+        public void TrimExcess()
+        {
+            TrimExcess(HashHelpers.GetPrime(Count));
```
Comment (jkotas) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: `GetPrime` does not need to be called here now that `TrimExcess(int capacity)` calls it.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 14]
```diff
@@ -777,6 +776,76 @@ public int EnsureCapacity(int capacity)
...
+                        freeList = entriesRef[index].next;
+                        ref Entry entry = ref entriesRef[i];
+                        entriesRef[index] = entry;
+                    }
+                }
+
+                _freeList = freeList;
+                _freeCount -= count - newSize;
+                _count -= count - newSize;
+                _version++;
```
Comment (jkotas) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Why to increment version? EnsureCapacity does not increment version either.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 15]
```diff
@@ -777,6 +776,76 @@ public int EnsureCapacity(int capacity)
             return newSize;
         }
 
+        /// <summary>
+        /// Sets the capacity of this dictionary to what it would be if it had been originally initialized with all its entries
+        /// 
+        /// This method can be used to minimize the memory overhead 
+        /// once it is known that no new elements will be added. 
+        /// 
+        /// To completely clear a dictionary and release all memory
```
Comment (jkotas) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Nit: This sequence will not release all memory referenced by the dictionary. It will allocate minimum size storage array.

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 16]
```diff
@@ -777,6 +776,76 @@ public int EnsureCapacity(int capacity)
...
+        {
+            if (capacity < Count)
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity);
+            int newSize = HashHelpers.GetPrime(capacity);
+            int currentCapacity = _entries == null ? 0 : _entries.Length;
+            if (newSize >= currentCapacity)
+                return;
+            if (newSize < _count)
+            {
+                ref Entry[] entriesRef = ref _entries;
```
Comment (jkotas) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: `ref` does not help performance here, quite the opposite. `ref` helps for the entries itself only .

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 17]
```diff
@@ -777,6 +776,76 @@ public int EnsureCapacity(int capacity)
...
+                    next = entriesRef[index].next;
+                    if (index < newSize)
+                    {
+                        entriesRef[index].next = freeList;
+                        freeList = index;
+                    }
+                    index = next;
+                }
+
+                int count = _count;
```
Comment (jkotas) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Just by looking at number of lines in this implementation, it is hard to believe that this would be the leanest way to implement this.

I have tried to write the most straightforward implementation possible: 
https://github.com/jkotas/coreclr/commit/ed51476d5d26a9be56220cfc1bb3b6b1500f2e16 It is smaller code than this.

Comment (maryamariyan) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: > @jkotas the original approach adds 19Kb to IL size. the implementation can be seen in commit 00c6b91
> A slight modification to original approach (dd514cd) would reduce 19Kb down to 18Kb, which is still not better than Approach2.

Your implementation here is the best one so far and as you said it adds less than others to native code (14.5Kb).
It's different from Approach2 (16Kb). I remember trying this (dd514cd) but since I was also doing an unnecessary Array.Copy prior to initialize, the code size ended up being higher (18Kb) and then I abandoned improving it further.

I'll update the PR to your approach right away.

Comment (danmosemsft) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: The purpose of version is to flag when enumeration is invalidated. That would happen iff the non-zombie entries in _entries are changed or re-ordered. So it seems to me that any operation which may do this, should version++.

I suppose there is value in flagging unsynchronized access to the object even if it does not change enumeration order but I'm guessing that isn't why version exists.

Comment (maryamariyan) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: I see now. In this case then I think it would have still been required in Approach2. Thanks for your explanation @danmosemsft

**Comment by maryamariyan** [comment]
@jkotas I addressed your latest feedback. 
Would you please take another look and let me know if you approve of the latest changes?

File: src/mscorlib/shared/System/Collections/Generic/Dictionary.cs [hunk 18]
```diff
@@ -512,8 +512,7 @@ private void Resize()
 
         private void Resize(int newSize, bool forceNewHashCodes)
         {
-            Debug.Assert(newSize >= _entries.Length);
-
+            Debug.Assert(newSize >= Count);
```
Comment (jkotas) [inline src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: Why did you change this assert?

Comment (maryamariyan) [reply src/mscorlib/shared/System/Collections/Generic/Dictionary.cs]: I'll revert it. This won't need to change since we are not calling Resize on TrimExcess anymore
